## Segmenting and clustering Toronto Neighborhoods.
### Get data about Toronto neighbourhoods
#### Step 1 - scrape Wikipedia page

In [1]:
import pandas as pd
url="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

tables = pd.read_html(url) #read html-page
df=tables[0] #create dataframe
df=df.loc[df['Borough']!='Not assigned'] #drop 'Not assigned' rows
df=df.reset_index(drop=True) #reset index and drop old index
df=df.rename(columns={'Postcode':'Postal Code'}) #rename column
df['Neighbourhood'].replace('Not assigned',df['Borough'],inplace=True) #Replace 'Not assigned' values in 'Neighbourhood' with 'Borough' values
df=df.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(', '.join).reset_index() #Group Neighbourhoods under same Boroughs
df.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [2]:
print('There are {} postal codes in {} boroughs in Toronto'.format(df['Postal Code'].unique().shape[0],df['Borough'].unique().shape[0]))


There are 103 postal codes in 11 boroughs in Toronto


#### Step2 - get latitude and longitude for Postal codes
#### Skip this step: geocoder is unstable.

In [3]:
# !pip install geocoder #install package

In [4]:
# import geocoder

# #Run loop of geocoder searches
# list_lat=[]
# list_long=[]
# for i, row in df.iterrows():
#     lat_lng_coords = None # initialize your variable to None
#     while(lat_lng_coords is None): # loop until you get the coordinates
#         g = geocoder.google(row['Postal Code'])
#         lat_lng_coords = g.latlng
#     list_lat.append(lat_lng_coords[0])
#     list_long.append(lat_lng_coords[1])

In [5]:
# df['Latitude']=list_lat
# df['Longitude']=list_long

<b>Step 2a</b> - Add latitude and longitude from CSV file

In [27]:
# The code was removed by Watson Studio for sharing.

In [28]:
df_data_1 = pd.read_csv(body)
df_data_1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [7]:
#df_data_1 - data read from Geospatial_coordinates.csv provided by the tutor
#The cell where df_data_1 defined is hidden as it contatins credentials to Cloud Storage
df=pd.merge(df,df_data_1,how='left', on='Postal Code') #Use 'left' merge get latitude and longitude from the Geospatial_coordinates.csv based on 'Postal Code'
df.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [8]:
# The code was removed by Watson Studio for sharing.

2019-10-25 19:57:56,832 - __PROJECT_LIB__ - ERROR - ProjectHandle: Access token neither provided nor found in the environment.


RuntimeError: Access token not available.

<b>Step 3. Import additional libraries for exploration and clustering</b>

In [8]:
import numpy as np # library to handle data in a vectorized manner
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library
print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /opt/conda/envs/Python36

  added / updated specs: 
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    openssl-1.1.1d             |       h516909a_0         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    altair-3.2.0               |           py36_0         770 KB  conda-forge
    certifi-2019.9.11          |           py36_0         147 KB  conda-forge
    ca-certificates-2019.9.11  |       hecc5488_0         144 KB  conda-forge
    folium-0.5.0               |             py_0          45 KB  conda-forge
    branca-0.3.1               |             py_0          25 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         3.2 MB

The following NEW packages will be 

<b>Step 4. Visualize postal codes on a map</b>

In [9]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[df['Latitude'].mean(), df['Longitude'].mean()], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Get data from Foursqure about Venues
#### Step 1. Connect to Foursquare.

In [11]:
from IPython.display import HTML
from IPython.display import display

In [ ]:
# The code was removed by Watson Studio for sharing.

In [12]:
neighborhood_latitude = df.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = df.loc[0, 'Postal Code'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of M1B are 43.806686299999996, -79.19435340000001.


#### Step 2. Create a function to get results from Foursquare. Create a dataframe by using this function

In [13]:
def getNearbyVenues(names, latitudes, longitudes, radius=500, LIMIT=100):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [14]:
Toronto_venues = getNearbyVenues(names=df['Neighbourhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

#### Step 3. Examine the new dataframe

In [15]:
Toronto_venues.groupby('Neighbourhood').count().sort_values('Venue',ascending=False)

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                               100   
Harbourfront East, Toronto Islands, Union Station                      100   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
St. James Town                                                         100   
Ryerson, Garden District                                               100   
First Canadian Place, Underground city                                 100   
Stn A PO Boxes 25 The Esplanade                                         99   
Chinatown, Grange Park, Kensington Market                               96   
Church and Wellesley                                                    88   
Central Bay Street                                                      80   
Fairview, Henry Farm, Oriole                                            65   
Little Portugal, Trinity                                                64   
Berczy Park                                                             55   
Harbourfront, Regent Park                                               50   
Cabbagetown, St. James Town                                             43   
Queen's Park                                                            42   
The Danforth West, Riverdale                                            42   
Studio District                                                         39   
Harbord, University of Toronto                                          37   
Willowdale South                                                        36   
Runnymede, Swansea                                                      35   
Leaside                                                                 33   
Davisville                                                              33   
High Park, The Junction South                                           25   
Flemingdon Park, Don Mills South                                        22   
Bedford Park, Lawrence Manor East                                       21   
Brockton, Exhibition Place, Parkdale Village                            21   
North Toronto West                                                      20   
The Annex, North Midtown, Yorkville                                     20   
...                                                                    ...   
The Beaches                                                              4   
Kingsview Village, Martin Grove Gardens, Richvi...                       4   
Agincourt                                                                4   
Caledonia-Fairbanks                                                      4   
Birch Cliff, Cliffside West                                              4   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                           4   
East Toronto                                                             4   
Humewood-Cedarvale                                                       4   
CFB Toronto, Downsview East                                              4   
Hillcrest Village                                                        4   
Humber Summit                                                            3   
Downsview, North Park, Upwood Park                                       3   
York Mills West                                                          3   
Lawrence Park                                                            3   
Humber Bay, King's Mill Park, Kingsway Park Sou...                       2   
The Kingsway, Montgomery Road, Old Mill North                            2   
Cliffcrest, Cliffside, Scarborough Village West                          2   
Moore Park, Summerhill East                                              2   
Downsview Central                                                        2

We can see that non-empty results were returned for 99 out of total 103 neighbourhoods.

In [16]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 272 uniques categories.


#### Step 4. Analyze neighbourhoods. Create "onehot" dataframe - For each neighbourhood calculate number of venues of each type

In [17]:
# one hot encoding - use get_dummies method from pandas
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Accessories Store  \
0                          Rouge, Malvern                  0   
1  Highland Creek, Rouge Hill, Port Union                  0   
2       Guildwood, Morningside, West Hill                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   ...  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0  ...                              0                 0            0   
1  ...                              0                 0            0   
2  ...                              0                 0            0   
3  ...                              0                 0            0   
4  ...                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0                0         0          0            0   
1                      0                0         0          0            0   
2                      0                0         0          0            0   
3                      0                0         0          0            0   
4                      0                0         0          0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 273 columns]

#### Step 5. Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [18]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Accessories Store  \
0                            Adelaide, King, Richmond                0.0   
1                                           Agincourt                0.0   
2   Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3   Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                              Alderwood, Long Branch                0.0   
5     Bathurst Manor, Downsview North, Wilson Heights                0.0   
6                                     Bayview Village                0.0   
7                   Bedford Park, Lawrence Manor East                0.0   
8                                         Berczy Park                0.0   
9                         Birch Cliff, Cliffside West                0.0   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...                0.0   
11       Brockton, Exhibition Place, Parkdale Village                0.0   
12  Business Reply Mail Processing Centre 969 Eastern                0.0   
13                        CFB Toronto, Downsview East                0.0   
14  CN Tower, Bathurst Quay, Island airport, Harbo...                0.0   
15                        Cabbagetown, St. James Town                0.0   
16                                Caledonia-Fairbanks                0.0   
17              Canada Post Gateway Processing Centre                0.0   
18                                          Cedarbrae                0.0   
19                                 Central Bay Street                0.0   
20          Chinatown, Grange Park, Kensington Market                0.0   
21                                           Christie                0.0   
22                               Church and Wellesley                0.0   
23                    Clairlea, Golden Mile, Oakridge                0.0   
24            Clarks Corners, Sullivan, Tam O'Shanter                0.0   
25    Cliffcrest, Cliffside, Scarborough Village West                0.0   
26  Cloverdale, Islington, Martin Grove, Princess ...                0.0   
27                     Commerce Court, Victoria Hotel                0.0   
28                                         Davisville                0.0   
29                                   Davisville North                0.0   
..                                                ...                ...   
69                                          Northwest                0.0   
70                    Northwood Park, York University                0.0   
71                             Parkdale, Roncesvalles                0.0   
72                                          Parkwoods                0.0   
73                                       Queen's Park                0.0   
74                                           Rosedale                0.0   
75                                           Roselawn                0.0   
76                                     Rouge, Malvern                0.0   
77                                 Runnymede, Swansea                0.0   
78                           Ryerson, Garden District                0.0   
79                                Scarborough Village                0.0   
80                           Silver Hills, York Mills                0.0   
81                                     St. James Town                0.0   
82                    Stn A PO Boxes 25 The Esplanade                0.0   
83                                    Studio District                0.0   
84                The Annex, North Midtown, Yorkville                0.0   
85                                        The Beaches                0.0   
86                     The Beaches West, India Bazaar                0.0   
87                       The Danforth West, Riverdale                0.0   
88                      The Junction North, Runnymede                0.0   
89      The Kingsway, Montgomery Road, Old Mill North                0.0   
90          

#### Step 6. Print top 5 venue categories for each neighbourhood

In [19]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3       Steakhouse  0.04
4  Thai Restaurant  0.03


----Agincourt----
                       venue  freq
0             Breakfast Spot  0.25
1                     Lounge  0.25
2  Latin American Restaurant  0.25
3               Skating Rink  0.25
4         Mexican Restaurant  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                venue  freq
0          Playground   0.5
1                Park   0.5
2  Miscellaneous Shop   0.0
3       Movie Theater   0.0
4               Motel   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0              Pharmacy  0.11
1         Grocery Store  0.11
2           Pizza Place  0.11
3  Fast Food Restaurant  0.11
4            Beer Store  0.11


----Alderwood, Long Branch----
                ven

#### Step 7. Create function to sort venues in descending order

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Step 8. Display top 10 venues for each neighbourhood

In [22]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...              Pharmacy   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0                  Café                   Bar                   Steakhouse   
1                Lounge          Skating Rink    Latin American Restaurant   
2                  Park           Yoga Studio  Eastern European Restaurant   
3            Beer Store  Fast Food Restaurant          Japanese Restaurant   
4              Pharmacy                   Pub                          Gym   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0   American Restaurant                Bakery      Sushi Restaurant   
1           Yoga Studio     Electronics Store            Donut Shop   
2        Discount Store               Dog Run      Doner Restaurant   
3        Sandwich Place        Discount Store   Fried Chicken Joint   
4          Skating Rink          Dance Studio           Coffee Shop   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0      Asian Restaurant       Thai Restaurant                   Restaurant  
1             Drugstore   Dumpling Restaurant  Eastern European Restaurant  
2            Donut Shop             Drugstore          Dumpling Restaurant  
3         Grocery Store           Pizza Place          Empanada Restaurant  
4        Sandwich Place                  Pool           Athletics & Sports

 ### Clustering
 #### Step 1. Run k-means to cluster the neighbourhood into 5 clusters

In [23]:
 # set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

#### Step 2. Add clustering labels and add it to the general dataframe

In [24]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels3', kmeans.labels_)

Toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

Toronto_merged.head() # check the last columns!

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels3 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353              0.0  Fast Food Restaurant           Yoga Studio   
1 -79.160497              4.0                   Bar           Yoga Studio   
2 -79.188711              0.0     Electronics Store                   Spa   
3 -79.216917              0.0           Coffee Shop     Convenience Store   
4 -79.239476              0.0  Caribbean Restaurant   Fried Chicken Joint   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Electronics Store               Dog Run      Doner Restaurant   
1     Electronics Store      Doner Restaurant            Donut Shop   
2           Pizza Place        Breakfast Spot          Intersection   
3     Korean Restaurant           Yoga Studio     Electronics Store   
4                  Bank       Thai Restaurant    Athletics & Sports   

  6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0            Donut Shop             Drugstore          Dumpling Restaurant   
1             Drugstore   Dumpling Restaurant  Eastern European Restaurant   
2        Medical Center   Rental Car Location           Mexican Restaurant   
3      Doner Restaurant            Donut Shop                    Drugstore   
4                Bakery                Lounge             Hakka Restaurant   

         9th Most Common Venue       10th Most Common Venue  
0  Eastern European Restaurant          Empanada Restaurant  
1          Empanada Restaurant   Financial or Legal Service  
2          Dumpling Restaurant             Doner Restaurant  
3          Dumpling Restaurant  Eastern European Restaurant  
4           Falafel Restaurant                  Event Space

#### Step 2a. Replace NA clusters as a new category

In [25]:
Toronto_merged_noNA = Toronto_merged.fillna(5)
Toronto_merged_noNA

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
..          ...               ...   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A      Queen's Park   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  \
0                                       Rouge, Malvern  43.806686 -79.194353   
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497   
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                               Woburn  43.770992 -79.216917   
4                                            Cedarbrae  43.773136 -79.239476   
5                                  Scarborough Village  43.744734 -79.239476   
6          East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029   
7                      Clairlea, Golden Mile, Oakridge  43.711112 -79.284577   
8      Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476   
9                          Birch Cliff, Cliffside West  43.692657 -79.264848   
10   Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304   
11                                   Maryvale, Wexford  43.750072 -79.295849   
12                                           Agincourt  43.794200 -79.262029   
13             Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302   
14   Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577   
15                                     L'Amoreaux West  43.799525 -79.318389   
16                                         Upper Rouge  43.836125 -79.205636   
17                                   Hillcrest Village  43.803762 -79.363452   
18                        Fairview, Henry Farm, Oriole  43.778517 -79.346556   
19                                     Bayview Village  43.786947 -79.385975   
20                           

#### Step 3. Visualize clustering with folium maps

In [26]:
# create map
map_clusters = folium.Map(location=[df['Latitude'].mean(), df['Longitude'].mean()], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged_noNA['Latitude'], Toronto_merged_noNA['Longitude'], Toronto_merged_noNA['Neighbourhood'], Toronto_merged_noNA['Cluster Labels3']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters